Fix block lock toolbar item stealing focus when mounted with StrictMode

jeryj opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
Potential fix to https://github.com/WordPress/gutenberg/issues/57139. I'd like to use BlockToolbar and `<NavigableToolbar />` instead, but haven't looked into if this is possible. This is just a quick patch with the current implementation.

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes block toolbar stealing focus on mount when using `<StrictMode />` due to `useEffect` running twice.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Fixing a bug exposed during `<StrictMode />` testing.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Adds a condition to only send focus to the toolbar if the modal has been opened while the toolbar is mounted. This fixes the main issue and makes it unlikely to steal focus unexpectedly.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
- Wrap the [`<PrivateBlockToolbar />` from `<BlockToolbarPopover />`](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-tools/block-toolbar-popover.js#L69-L80) in a `<StrictMode />` component (`import { StrictMode } from 'react';`).
- Add a paragraph block
- Type
- Move the mouse to trigger the block toolbar popover to show
- Focus should not be placed on the toolbar
- Lock the block
- From the Lock icon in the toolbar, trigger the modal
- Unlock the block
- Focus should be sent to the toolbar

## Screenshots or screencast <!-- if applicable -->

jeryj committed:
Fix block lock toolbar item stealing focus when mounted with StrictMode

youknowriad commented on packages/block-editor/src/components/block-lock/toolbar.js:
```diff
@@ -33,10 +34,23 @@ export default function BlockLockToolbar( { clientId, wrapperRef } ) {
+		// We only want to allow this effeect to happen if the modal has been opened.
+		// The issue is when we're returning focus from the block lock modal to a toolbar,
+		// so it can only happen after a modal has been opened. Without this, the toolbar
+		// will steal focus on rerenders.
```
Based on this comment, I'm wondering if this fix will cover all use-cases. What happens if the modal has been opened and closed, could this effect trigger at another moment after the modal was closed (not right after).

jeryj replied:
I agree this isn't a great fix. Definitely a hotfix at best. It's better than the current check, since it at least only surfaces in a rarer instance than the current check.

Rather than try to guess if we need to force focus to the toolbar (and where to focus), I think a better solution would be:

**If the Lock Button has rendered, force it to stay in the toolbar until it is unmounted as a "lock" action button.**

This would mean there would be an "Unlocked" icon until the block toolbar is has been unmounted, but it would improve the UX since the modal could always send focus back to the button that triggered it (and fix this focus stealing issue). Clicking the "Unlocked" icon would trigger the lock modal again. Again, this would only show _after unlocking_ the block. It would not show for all toolbars.

<img width="502" alt="Block toolbar with an 'unlocked' icon" src="https://github.com/WordPress/gutenberg/assets/967608/512608b6-d272-471d-89f5-9e557e09dec2">

ciampo replied:
Agreed — this sounds like a good idea that we should try out separately from this hot fix.

youknowriad reviewed (approved):
Would love a look from Marco here, but this I'm fine shipping this hot fix. Thanks Jerry for looking into it.

ciampo commented on packages/block-editor/src/components/block-lock/toolbar.js:
```diff
@@ -33,10 +34,23 @@ export default function BlockLockToolbar( { clientId, wrapperRef } ) {
 	useEffect( () => {
 		if ( isFirstRender.current ) {
 			isFirstRender.current = false;
+			hasModalOpened.current = false;
```
Since `hasModalOpened.current` is already initialized as `false` , is this line necessary?

ciampo commented on packages/block-editor/src/components/block-lock/toolbar.js:
```diff
@@ -49,7 +63,7 @@ export default function BlockLockToolbar( { clientId, wrapperRef } ) {
 		}
 		// wrapperRef is a reference object and should be stable
-	}, [ isModalOpen, shouldHideBlockLockUI, wrapperRef ] );
+	}, [ isModalOpen, shouldHideBlockLockUI, hasModalOpened, wrapperRef ] );
```
Since `hasModalOpened` is a ref, I don't think it should be listed as a `useEffect` dependency

ciampo commented on packages/block-editor/src/components/block-lock/toolbar.js:
```diff
@@ -33,10 +34,23 @@ export default function BlockLockToolbar( { clientId, wrapperRef } ) {
+			hasModalOpened.current = true;
+		}
+
+		// We only want to allow this effeect to happen if the modal has been opened.
```
Nit: typo

```suggestion
		// We only want to allow this effect to happen if the modal has been opened.
```

ciampo reviewed (commented):
Thank you for the ping, and thank you for iterating on this fix!

I left a few suggestions, although I'm not sure how to test them in the editor and/or Storybook (I didn't know where to look for testing instructions)

ciampo commented on packages/block-editor/src/components/block-lock/toolbar.js:
```diff
@@ -33,10 +34,23 @@ export default function BlockLockToolbar( { clientId, wrapperRef } ) {
 		}
 
-		if ( ! isModalOpen && shouldHideBlockLockUI ) {
+		if ( isModalOpen && ! hasModalOpened.current ) {
+			hasModalOpened.current = true;
+		}
```
(Optional) Putting together a couple of suggestions left above, we could also split the logic into three smaller `useEffect`s — one for the "first render" logic, one for the "has modal ever opened" logic, and one for focus restoration

```diff
diff --git a/packages/block-editor/src/components/block-lock/toolbar.js b/packages/block-editor/src/components/block-lock/toolbar.js
index 3385dd1b15..2c69fe3143 100644
--- a/packages/block-editor/src/components/block-lock/toolbar.js
+++ b/packages/block-editor/src/components/block-lock/toolbar.js
@@ -28,42 +28,47 @@ export default function BlockLockToolbar( { clientId, wrapperRef } ) {
 	const shouldHideBlockLockUI =
 		! canLock || ( canEdit && canMove && canRemove );
 
-	// Restore focus manually on the first focusable element in the toolbar
-	// when the block lock modal is closed and the block is not locked anymore.
-	// See https://github.com/WordPress/gutenberg/issues/51447
 	useEffect( () => {
 		if ( isFirstRender.current ) {
 			isFirstRender.current = false;
-			hasModalOpened.current = false;
-			return;
 		}
+	}, [] );
 
+	useEffect( () => {
 		if ( isModalOpen && ! hasModalOpened.current ) {
 			hasModalOpened.current = true;
 		}
+	}, [ isModalOpen ] );
 
+	// Restore focus manually on the first focusable element in the toolbar
+	// when the block lock modal is closed and the block is not locked anymore.
+	// See https://github.com/WordPress/gutenberg/issues/51447
+	useEffect( () => {
 		// We only want to allow this effeect to happen if the modal has been opened.
 		// The issue is when we're returning focus from the block lock modal to a toolbar,
 		// so it can only happen after a modal has been opened. Without this, the toolbar
 		// will steal focus on rerenders.
 		if (
-			hasModalOpened.current &&
-			! isModalOpen &&
-			shouldHideBlockLockUI
+			isFirstRender.current ||
+			! hasModalOpened.current ||
+			isModalOpen ||
+			! shouldHideBlockLockUI
 		) {
-			focus.focusable
-				.find( wrapperRef.current, {
-					sequential: false,
-				} )
-				.find(
-					( element ) =>
-						element.tagName === 'BUTTON' &&
-						element !== lockButtonRef.current
-				)
-				?.focus();
+			return;
 		}
+
+		focus.focusable
+			.find( wrapperRef.current, {
+				sequential: false,
+			} )
+			.find(
+				( element ) =>
+					element.tagName === 'BUTTON' &&
+					element !== lockButtonRef.current
+			)
+			?.focus();
 		// wrapperRef is a reference object and should be stable
-	}, [ isModalOpen, shouldHideBlockLockUI, hasModalOpened, wrapperRef ] );
+	}, [ isModalOpen, shouldHideBlockLockUI, wrapperRef ] );
 
 	if ( shouldHideBlockLockUI ) {
 		return null;

```

jeryj committed:
Address feedback

jeryj commented:
@ciampo Thanks for the review! I addressed the feedback other than restructuring the `useEffect`, as I'd like to move away from this set-up anyways if possible. I have an alternative fix at https://github.com/WordPress/gutenberg/pull/57229/ ready for review as well.
